bpo-41930: Add support for SQLite serialise/deserialise API

erlend-aasland opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-41930](https://bugs.python.org/issue41930) -->
https://bugs.python.org/issue41930
<!-- /issue-number -->

Commit:
Expose sqlite3_serialize and sqlite3_deserialize

Commit:
Wrap in ifdef for now

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Add basic tests

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Let AC convert data to buffer

Commit:
Test deserialize type errors

Commit:
Skip tests if serialize API is missing

Commit:
Cast sqlite3_deserialize() arguments

Commit:
Only reset pending statements if needed

Commit:
Refactor tests

Commit:
Check thread and connection when deserializing

Commit:
Test deserialize with corrupt data

Commit:
Simplify serialize return value

Don't return size; use len() on the returned bytes object instead.

Commit:
Reset cursors and statements before deserialize

Add basic test for this

Commit:
Add NEWS stub

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Regen clinic after GH-27155.

Commit:
Merge branch 'main' into sqlite-serialize

Sync with main bco. GH-27273

Commit:
Free SQLite buffer after Py conversion

Commit:
Transfer ownership of the deserialized data to SQLite

Commit:
Add docstrings

Commit:
Merge branch 'main' into sqlite-serialize

Sync with main bco. GH-27431

Commit:
Add What's New entry

Commit:
Add Docs

Commit:
Improve naming: schema => name

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Add more tests

Commit:
Allow threads while serializing

Commit:
skip test if SQLite version < 3.36.0

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Regen clinic

Commit:
Fix merge

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Improve tests

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Use autoconf to detect serialize API

Commit:
Regen clinic

Commit:
Fix test deps

Commit:
Build sqlite3 with serialize on Windows

Commit:
Typo

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Fix skipIf decorator

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Note that these API's are only available if the underlying SQLite lib supports them

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Fix merge: pysqlite_do_all_statements is an ex-function

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Use sqlite3_malloc64 and add bigmemtest

Commit:
Remove spurious newline in VS docs

Commit:
Merge branch 'main' into sqlite-serialize

Commit:
Nit: add missing punctuation

Commit:
Allow threads while memcpy'ing

Commit:
Remove extra newline

Commit:
Regen clinic

JelleZijlstra commented on Doc/library/sqlite3.rst:
```diff
@@ -748,6 +748,41 @@ Connection Objects
 
+   .. method:: serialize(*, name="main")
+
+      This method serializes a database into a :class:`bytes` sequence.  For an
```
```suggestion
      This method serializes a database into a :class:`bytes` object.  For an
```
Your wording sounds like it returns a list of bytes objects.

erlend-aasland replied:
Thanks!

JelleZijlstra commented on Doc/library/sqlite3.rst:
```diff
@@ -748,6 +748,41 @@ Connection Objects
+      ordinary on-disk database file, the serialization is just a copy of the
+      disk file.  For an in-memory database or a "temp" database, the
+      serialization is the same sequence of bytes which would be written to
+      disk if that database where backed up to disk.
```
```suggestion
      disk if that database were backed up to disk.
```

JelleZijlstra commented on Modules/_sqlite/connection.c:
```diff
@@ -1817,6 +1817,117 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+For an ordinary on-disk database file, the serialization is just a copy of the
+disk file. For an in-memory database or a "temp" database, the serialization is
+the same sequence of bytes which would be written to disk if that database
+where backed up to disk.
```
```suggestion
were backed up to disk.
```

JelleZijlstra commented on Modules/_sqlite/connection.c:
```diff
@@ -1817,6 +1817,117 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+     * We can safely use sqlite3_malloc64 here, since it was introduced before
+     * the serialize APIs.
+     */
+    if (data->len > 9223372036854775807) {
```
```suggestion
    if (data->len >= (2 << 62)) {
```
Or some other variation that's not just a magic number.

erlend-aasland replied:
I added a comment instead; it seems that the MSVC compiler does some tricks here (I tried to find out via godbolt.org, but to no avail; perhaps explicitly marking the shifted literal as `long long` would have worked...) The code base also seem to prefer the explicit to the implicit here:

```
$ grep -r "922337203685477580[78]" * | wc -l
      47
$ grep -r "[12] << 6[23]" * | wc -l          
      13
```

erlend-aasland replied:
On MSVC, `(1 << 63) - 1` triggers [MSVC compiler warning C4293](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4293?view=msvc-170). Using `(1LL << 63) - 1` seems to fix it, but the MS docs promotes casting to `unsigned __int64`.

I prefer just writing it out instead; also, there is already a comment explaining why we do this :)

```
     * Make sure we don't overflow sqlite3_deserialize(); it accepts a signed
     * 64-bit int as its data size argument, but sqlite3_malloc64 accepts an
     * unsigned 64-bit int as its size argument.
```

erlend-aasland replied:
(Actually, I can reduce that comment; only the first part is important.)

JelleZijlstra commented on pyconfig.h.in:
```diff
@@ -1494,6 +1494,9 @@
 /* Define to 1 to build the sqlite module with loadable extensions support. */
 #undef PY_SQLITE_ENABLE_LOAD_EXTENSION
 
+/* "Define if SQLite was compiled with the serialize API" */
```
```suggestion
/* Define if SQLite was compiled with the serialize API */
```

erlend-aasland replied:
FYI, this is generated by `AC_DEFINE` in `configure.ac`.

JelleZijlstra reviewed (commented):
Took a first look and noticed some small things.

Commit:
Update Doc/library/sqlite3.rst

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

Commit:
Address code review

Commit:
Revert int max change

Commit:
Remove redundant info from comment

JelleZijlstra commented on Modules/_sqlite/connection.c:
```diff
@@ -1817,6 +1817,116 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+    const char *data;
+
+    Py_BEGIN_ALLOW_THREADS
+    data = (const char *)sqlite3_serialize(self->db, name, &size, flags);
```
reference: https://www.sqlite.org/c3ref/serialize.html

It sounds like the `SQLITE_SERIALIZE_NOCOPY` could provide a useful optimization, since we end up immediately copying the memory into a new bytes object. However, it would be more complex to use; most likely we'd have to first try with `SQLITE_SERIALIZE_NOCOPY` on, then if that fails, try again with it off. I don't know if that's worth the complexity, and in any case we're free to change it later if a user complains about the speed of repeatedly serializing a database.

erlend-aasland replied:
Let's try!

erlend-aasland replied:
See 7edccfed418eca5cdf88517cf9f3e91c2a051bc8.

JelleZijlstra commented on Modules/_sqlite/connection.c:
```diff
@@ -1817,6 +1817,116 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+    name: str = "main"
+        Which database to serialize.
+
+Serialize a database into a byte string. Non-standard.
```
What does "non-standard" mean here? I see it on a lot of other docstrings in this file, but SQLite's own documentation doesn't say anything similar.

erlend-aasland replied:
It's a historic artefact; Gerhard marked all non-PEP-249 methods with "non-standard". Let's just remove those in another PR; I know it is confusing to some (most?) people.

erlend-aasland replied:
See GH-31612

JelleZijlstra commented on Modules/_sqlite/connection.c:
```diff
@@ -1817,6 +1817,116 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+    int rc;
+    Py_BEGIN_ALLOW_THREADS
+    (void)memcpy(buf, data->buf, data->len);
+    rc = sqlite3_deserialize(self->db, name, buf, size, size, flags);
```
I had to convince myself the memory handling here was right. Spoiler alert: it is. Reference https://www.sqlite.org/c3ref/deserialize.html.

We set the FREEONCLOSE flag, which means that sqlite will treat the data buffer as its own, and free it when the database is closed. It will also free it when the function fails, so it's correct that the error case on line 1921  below doesn't free anything.

(No changes requested)

erlend-aasland replied:
Yeah, it took me a little while to get that right; I misread the docs first, so the first iteration of this PR had it all wrong 😆

JelleZijlstra commented on Modules/_sqlite/connection.c:
```diff
@@ -1817,6 +1817,116 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+        (void)_pysqlite_seterror(self->state, self->db);
+        return NULL;
+    }
+    Py_RETURN_TRUE;
```
Should we return None instead since the function never returns False?

erlend-aasland replied:
Yes, we should.

JelleZijlstra commented on Doc/library/sqlite3.rst:
```diff
@@ -748,6 +748,41 @@ Connection Objects
+      .. note::
+
+         This method is only available if the underlying SQLite library has the
+         serialize API.
```
```suggestion
         deserialize API.
```

JelleZijlstra commented on Doc/library/sqlite3.rst:
```diff
@@ -748,6 +748,41 @@ Connection Objects
+      This method causes the database connection to disconnect from database
+      *name*, and reopen *name* as an in-memory database based on the
+      serialization contained in *data*.  Deserialization will fail with
+      ``SQLITE_BUSY`` if the database is currently in a read transaction or is
```
That's true at the C level, but for the Python-level documentation it will be more useful to say what exception it will throw.

Generally I noticed that the documentation is partially copied from SQLite's own docs. That should be fine because SQLite is in the public domain and obviously it's a good source for documentation about SQLite itself.

erlend-aasland replied:
> Generally I noticed that the documentation is partially copied from SQLite's own docs. That should be fine because SQLite is in the public domain and obviously it's a good source for documentation about SQLite itself.

Yes, your observation is correct.

> That's true at the C level, but for the Python-level documentation it will be more useful to say what exception it will throw.

Yes, I'll reword it.

JelleZijlstra reviewed (changes requested):
Thanks. I'm on board with merging this, since it's generally useful for sqlite3 to expose more of the underlying SQLite API, the feature is simple and shouldn't need much maintenance in the future, and there haven't been any objections on the issue.

I do have a few small comments. Once those are addressed, I'll ask Guido for a second look and then merge the PR.

Commit:
Update Doc/library/sqlite3.rst

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

Commit:
Address review

- adjust docs
- deserialize returns None
- remove 'non-standard' from docstring

Commit:
Address review: try to avoid memory allocations when serializing

JelleZijlstra reviewed (approved):
Looks good! @gvanrossum I'm planning to merge this PR (adding two new functions to sqlite3)

kumaraditya303 commented on PCbuild/_sqlite3.vcxproj:
```diff
@@ -94,6 +94,7 @@
   <ItemDefinitionGroup>
     <ClCompile>
       <AdditionalIncludeDirectories>$(sqlite3Dir);%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <PreprocessorDefinitions>PY_SQLITE_HAVE_SERIALIZE</PreprocessorDefinitions>
```
```suggestion
      <PreprocessorDefinitions>PY_SQLITE_HAVE_SERIALIZE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```

erlend-aasland replied:
Thanks!

Commit:
Update PCbuild/_sqlite3.vcxproj

Co-authored-by: Kumar Aditya <59607654+kumaraditya303@users.noreply.github.com>

Commit:
Merge branch 'main' and regen clinic

Commit:
Merge branch 'main' into sqlite-serialize
